File: src/text_utils.ts
import type { Node, BlockNode } from "./ast.ts";

export type TextRegion = "deleted" | "inserted" | "current";

/**
 * The private-note type of an invisible-in-output note inset block — `Note`
 * (Note Note) or `Comment` (Note Comment) — or undefined for anything else
 * (Greyedout is visible output and is NOT included).
 */
export function invisibleInsetType(block: BlockNode): string | undefined {
  if (block.type !== "block" || block.tag !== "inset") return undefined;
  const args = (block.args ?? "").trim();
  if (args === "Note Note") return "Note";
  if (args === "Note Comment") return "Comment";
  return undefined;
}

/**
 * Is this block a private (invisible-in-output) note inset — `Note Note` or
 * `Note Comment`? `Note Greyedout` is visible output and is NOT included.
 * Shared by the query engine and split-after so the private set lives in one
 * place (dev log 99 — F2 notes visibility).
 */
export function isInvisibleInset(block: BlockNode): boolean {
  return invisibleInsetType(block) !== undefined;
}

export interface TraversalState {
  deletedDepth: number;
  insertedDepth: number;
  deletedAuthor: number;
  deletedTs: string;
  insertedAuthor: number;
  insertedTs: string;
  outerDeletedDepth: number;
  outerInsertedDepth: number;
  outerDeletedAuthor: number;
  outerDeletedTs: string;
  outerInsertedAuthor: number;
  outerInsertedTs: string;
  properties: Record<string, string | undefined>;
}

export function createTraversalState(): TraversalState {
  return {
    deletedDepth: 0,
    insertedDepth: 0,
    deletedAuthor: 0,
    deletedTs: "",
    insertedAuthor: 0,
    insertedTs: "",
    outerDeletedDepth: 0,
    outerInsertedDepth: 0,
    outerDeletedAuthor: 0,
    outerDeletedTs: "",
    outerInsertedAuthor: 0,
    outerInsertedTs: "",
    properties: {},
  };
}

export function cloneTraversalState(state: TraversalState): TraversalState {
  return {
    ...state,
    properties: { ...state.properties },
  };
}

export function traversalRegion(state: TraversalState): TextRegion {
  if (state.deletedDepth > 0) return "deleted";
  if (state.insertedDepth > 0) return "inserted";
  if (state.outerDeletedDepth > 0) return "deleted";
  if (state.outerInsertedDepth > 0) return "inserted";
  return "current";
}

export function traversalChange(state: TraversalState): { author: number; ts: string } {
  if (state.deletedDepth > 0) return { author: state.deletedAuthor, ts: state.deletedTs };
  if (state.insertedDepth > 0) return { author: state.insertedAuthor, ts: state.insertedTs };
  if (state.outerDeletedDepth > 0) return { author: state.outerDeletedAuthor, ts: state.outerDeletedTs };
  if (state.outerInsertedDepth > 0) return { author: state.outerInsertedAuthor, ts: state.outerInsertedTs };
  return { author: 0, ts: "" };
}

/** Enter a nested structural child list without losing its enclosing state. */
export function enterTraversalState(parent: TraversalState): TraversalState {
  const child = createTraversalState();
  const region = traversalRegion(parent);
  const change = traversalChange(parent);
  if (region === "deleted") {
    child.outerDeletedDepth = 1;
    child.outerDeletedAuthor = change.author;
    child.outerDeletedTs = change.ts;
  } else if (region === "inserted") {
    child.outerInsertedDepth = 1;
    child.outerInsertedAuthor = change.author;
    child.outerInsertedTs = change.ts;
  }
  child.properties = { ...parent.properties };
  return child;
}

/** Apply one LyX property marker to an inherited recursive traversal state. */
export function advanceTraversalState(
  state: TraversalState,
  key: string,
  value?: string,
): void {
  if (key === "change_deleted" || key === "change_inserted" || key === "change_unchanged") {
    const previousDeletedDepth = state.deletedDepth;
    const previousInsertedDepth = state.insertedDepth;
    const depths = advanceChangeDepths(key, previousDeletedDepth, previousInsertedDepth);
    state.deletedDepth = depths.deletedDepth;
    state.insertedDepth = depths.insertedDepth;

    if (key === "change_deleted") {
      const parts = value?.trim().split(/\s+/) ?? [];
      state.deletedAuthor = parseInt(parts[0] ?? "", 10) || 0;
      state.deletedTs = parts[1] ?? "0";
      state.insertedAuthor = 0;
      state.insertedTs = "";
    } else if (key === "change_inserted") {
      const parts = value?.trim().split(/\s+/) ?? [];
      state.insertedAuthor = parseInt(parts[0] ?? "", 10) || 0;
      state.insertedTs = parts[1] ?? "0";
      state.deletedAuthor = 0;
      state.deletedTs = "";
    } else {
      if (previousDeletedDepth > 0 && state.deletedDepth === 0) {
        state.deletedAuthor = 0;
        state.deletedTs = "";
      }
      if (previousInsertedDepth > 0 && state.insertedDepth === 0) {
        state.insertedAuthor = 0;
        state.insertedTs = "";
      }
    }
    return;
  }

  state.properties[key] = value;
}

export interface TextSegment {
  /** Index of this text node in the children array that owns it */
  childIndex: number;
  /** The text content of this node */
  text: string;
  /** The children array containing this node at `childIndex` — set when
   *  `recurseLayouts` is used, so `split-after` can splice into the list the
   *  text actually lives in (a match may sit in a nested layout, e.g. inside a
   *  Foot inset's Plain Layout — test_report_38 F3). */
  owner?: Node[];
  /** Effective change/style state at this text node, including enclosing insets. */
  state: TraversalState;
}

/**
 * Advance LyX's flat-model change depths when a change marker is encountered.
 * LyX keeps one active Change per position (dev log 84 F1): a different-type
 * opener terminates any open region of the other type (the opposite depth is
 * zeroed), and a closer decrements whichever depth is open. Every marker
 * walker routes its depth bookkeeping through this single implementation so
 * the rule lives in one place (code_review_85-74 Standards-3).
 */
export function advanceChangeDepths(
  key: string,
  deletedDepth: number,
  insertedDepth: number,
): { deletedDepth: number; insertedDepth: number } {
  if (key === "change_deleted") return { deletedDepth: deletedDepth + 1, insertedDepth: 0 };
  if (key === "change_inserted") return { deletedDepth: 0, insertedDepth: insertedDepth + 1 };
  // \change_unchanged: close whichever region is open
  if (insertedDepth > 0) return { deletedDepth, insertedDepth: insertedDepth - 1 };
  if (deletedDepth > 0) return { deletedDepth: deletedDepth - 1, insertedDepth };
  return { deletedDepth, insertedDepth };
}

/**
 * Build a concatenated view of text children. Callers can exclude text inside
 * \change_deleted blocks with the default `includeDeleted: false`; selector
 * and mutation callers opt into the see-all view. Change tracking markers (change_deleted,
 * change_inserted, change_unchanged) are transparent — they don't
 * break concatenation. Only structural children (insets, nested
 * layouts) act as implicit boundaries (detected via childIndex gaps).
 *
 * With `opts.recurseLayouts`, text is also collected from nested `layout`
 * blocks (descending through `inset` blocks like `Foot` to reach them), so
 * `split-after` can target text inside a footnote (test_report_38 F3). Text is
 * collected only under a layout container — inset metadata (a Foot `status`
 * line, a CommandInset `name "…"` line) is not matchable. Enclosing change
 * and style state is inherited through those structural blocks. Each segment
 * carries its owning children list (`owner`) and effective state.
 */
export function concatenateTextNodes(
  children: Node[],
  opts?: {
    recurseLayouts?: boolean;
    topLevelIsLayout?: boolean;
    includeDeleted?: boolean;
    inheritedState?: TraversalState;
    /** DL99: when true (and recurseLayouts), do not descend into private note
     *  insets (Note Note / Note Comment) — split-after must not leak into them
     *  unless the target is note-scoped. Inert when recurseLayouts is false, so
     *  the --find / :contains non-recursive paths are unaffected. */
    skipInvisibleNotes?: boolean;
  },
): { segments: TextSegment[]; fullText: string } {
  const segments: TextSegment[] = [];
  const recurse = opts?.recurseLayouts ?? false;
  const topLevelIsLayout = opts?.topLevelIsLayout ?? true;
  const includeDeleted = opts?.includeDeleted ?? false;
  const skipInvisibleNotes = opts?.skipInvisibleNotes ?? false;

  function walk(list: Node[], collectText: boolean, inheritedState: TraversalState): void {
    const state = enterTraversalState(inheritedState);
    for (let i = 0; i < list.length; i++) {
      const child = list[i];
      if (child.type === "property") {
        advanceTraversalState(state, child.key, child.value);
      } else if (child.type === "text") {
        // Selectors and mutation callers pass includeDeleted=true so a phrase
        // inside \\change_deleted remains reachable; callers that need a
        // current-only view can leave the option false.
        if (collectText && (includeDeleted || traversalRegion(state) !== "deleted")) {
          const segmentState = cloneTraversalState(state);
          if (recurse) {
            segments.push({ childIndex: i, text: child.text, owner: list, state: segmentState });
          } else {
            segments.push({ childIndex: i, text: child.text, state: segmentState });
          }
        }
      } else if (child.type === "block" && recurse) {
        const b = child as BlockNode;
        if (b.tag === "layout") {
          walk(b.children, true, state); // collect text under nested layouts
        } else if (b.tag === "inset") {
          // DL99: private notes are invisible content — do not descend into
          // them when skipInvisibleNotes is set (split-after leak fix).
          if (skipInvisibleNotes && isInvisibleInset(b)) continue;
          // Descend through insets to reach their nested layouts, but do NOT
          // collect inset metadata text (e.g. a CommandInset label line).
          walk(b.children, false, state);
        }
      }
    }
  }
  walk(children, topLevelIsLayout, opts?.inheritedState ?? createTraversalState());

  const fullText = segments.map(s => s.text).join("");
  return { segments, fullText };
}

/** Map a position in the concatenated fullText to (segmentIndex, offsetInSegment). */
export function mapPosToSegment(segments: TextSegment[], pos: number): { segIdx: number; offset: number } {
  let remaining = pos;
  for (let i = 0; i < segments.length; i++) {
    if (remaining < segments[i].text.length) return { segIdx: i, offset: remaining };
    remaining -= segments[i].text.length;
  }
  const last = segments[segments.length - 1];
  return { segIdx: segments.length - 1, offset: last.text.length };
}

File: src/text_utils.rs
//! Flat-model text walkers (Deno `text_utils.ts`).

use crate::ast::{Document, NodeId, NodeKind};
use std::collections::HashMap;

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum TextRegion {
    Deleted,
    Inserted,
    Current,
}

pub fn invisible_inset_type(tag: &str, args: Option<&str>) -> Option<&'static str> {
    if tag != "inset" {
        return None;
    }
    match args.unwrap_or("").trim() {
        "Note Note" => Some("Note"),
        "Note Comment" => Some("Comment"),
        _ => None,
    }
}

pub fn is_invisible_inset(tag: &str, args: Option<&str>) -> bool {
    invisible_inset_type(tag, args).is_some()
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct TraversalState {
    pub deleted_depth: i32,
    pub inserted_depth: i32,
    pub deleted_author: i32,
    pub deleted_ts: String,
    pub inserted_author: i32,
    pub inserted_ts: String,
    pub outer_deleted_depth: i32,
    pub outer_inserted_depth: i32,
    pub outer_deleted_author: i32,
    pub outer_deleted_ts: String,
    pub outer_inserted_author: i32,
    pub outer_inserted_ts: String,
    pub properties: HashMap<String, Option<String>>,
}

pub(crate) fn create_traversal_state() -> TraversalState {
    TraversalState {
        deleted_depth: 0,
        inserted_depth: 0,
        deleted_author: 0,
        deleted_ts: String::new(),
        inserted_author: 0,
        inserted_ts: String::new(),
        outer_deleted_depth: 0,
        outer_inserted_depth: 0,
        outer_deleted_author: 0,
        outer_deleted_ts: String::new(),
        outer_inserted_author: 0,
        outer_inserted_ts: String::new(),
        properties: HashMap::new(),
    }
}

pub(crate) fn clone_traversal_state(state: &TraversalState) -> TraversalState {
    state.clone()
}

pub(crate) fn traversal_region(state: &TraversalState) -> TextRegion {
    if state.deleted_depth > 0 {
        TextRegion::Deleted
    } else if state.inserted_depth > 0 {
        TextRegion::Inserted
    } else if state.outer_deleted_depth > 0 {
        TextRegion::Deleted
    } else if state.outer_inserted_depth > 0 {
        TextRegion::Inserted
    } else {
        TextRegion::Current
    }
}

pub(crate) fn traversal_change(state: &TraversalState) -> (i32, String) {
    if state.deleted_depth > 0 {
        (state.deleted_author, state.deleted_ts.clone())
    } else if state.inserted_depth > 0 {
        (state.inserted_author, state.inserted_ts.clone())
    } else if state.outer_deleted_depth > 0 {
        (state.outer_deleted_author, state.outer_deleted_ts.clone())
    } else if state.outer_inserted_depth > 0 {
        (state.outer_inserted_author, state.outer_inserted_ts.clone())
    } else {
        (0, String::new())
    }
}

pub(crate) fn enter_traversal_state(parent: &TraversalState) -> TraversalState {
    let mut child = create_traversal_state();
    let region = traversal_region(parent);
    let (author, ts) = traversal_change(parent);
    match region {
        TextRegion::Deleted => {
            child.outer_deleted_depth = 1;
            child.outer_deleted_author = author;
            child.outer_deleted_ts = ts;
        }
        TextRegion::Inserted => {
            child.outer_inserted_depth = 1;
            child.outer_inserted_author = author;
            child.outer_inserted_ts = ts;
        }
        TextRegion::Current => {}
    }
    child.properties = parent.properties.clone();
    child
}

fn parse_author_id(s: &str) -> i32 {
    s.parse::<i32>().unwrap_or(0)
}

pub(crate) fn advance_traversal_state(state: &mut TraversalState, key: &str, value: Option<&str>) {
    if key == "change_deleted" || key == "change_inserted" || key == "change_unchanged" {
        let previous_deleted_depth = state.deleted_depth;
        let previous_inserted_depth = state.inserted_depth;
        let (deleted_depth, inserted_depth) =
            advance_change_depths(key, previous_deleted_depth, previous_inserted_depth);
        state.deleted_depth = deleted_depth;
        state.inserted_depth = inserted_depth;

        if key == "change_deleted" {
            let parts: Vec<&str> = value.unwrap_or("").split_whitespace().collect();
            state.deleted_author = parse_author_id(parts.first().copied().unwrap_or(""));
            state.deleted_ts = parts.get(1).unwrap_or(&"0").to_string();
            state.inserted_author = 0;
            state.inserted_ts.clear();
        } else if key == "change_inserted" {
            let parts: Vec<&str> = value.unwrap_or("").split_whitespace().collect();
            state.inserted_author = parse_author_id(parts.first().copied().unwrap_or(""));
            state.inserted_ts = parts.get(1).unwrap_or(&"0").to_string();
            state.deleted_author = 0;
            state.deleted_ts.clear();
        } else {
            if previous_deleted_depth > 0 && state.deleted_depth == 0 {
                state.deleted_author = 0;
                state.deleted_ts.clear();
            }
            if previous_inserted_depth > 0 && state.inserted_depth == 0 {
                state.inserted_author = 0;
                state.inserted_ts.clear();
            }
        }
        return;
    }

    state
        .properties
        .insert(key.to_string(), value.map(str::to_string));
}

#[derive(Clone, Debug)]
pub struct TextSegment {
    pub child_index: usize,
    pub text: String,
    pub owner: Option<NodeId>,
    pub state: TraversalState,
}

pub fn advance_change_depths(key: &str, deleted_depth: i32, inserted_depth: i32) -> (i32, i32) {
    if key == "change_deleted" {
        return (deleted_depth + 1, 0);
    }
    if key == "change_inserted" {
        return (0, inserted_depth + 1);
    }
    if inserted_depth > 0 {
        return (deleted_depth, inserted_depth - 1);
    }
    if deleted_depth > 0 {
        return (deleted_depth - 1, inserted_depth);
    }
    (deleted_depth, inserted_depth)
}

fn is_change_marker_key(key: &str) -> bool {
    key == "change_deleted" || key == "change_inserted" || key == "change_unchanged"
}

/// Region of `parent`'s child at `index` from sibling change markers only.
pub(crate) fn region_at(doc: &Document, parent: NodeId, index: usize) -> TextRegion {
    let mut deleted_depth = 0i32;
    let mut inserted_depth = 0i32;
    for &id in doc.node(parent).children.iter().take(index) {
        if let NodeKind::Property { key, .. } = &doc.node(id).kind
            && is_change_marker_key(key)
        {
            let (d, ins) = advance_change_depths(key, deleted_depth, inserted_depth);
            deleted_depth = d;
            inserted_depth = ins;
        }
    }
    if deleted_depth > 0 {
        TextRegion::Deleted
    } else if inserted_depth > 0 {
        TextRegion::Inserted
    } else {
        TextRegion::Current
    }
}

#[derive(Clone, Debug)]
pub struct ConcatOpts {
    pub recurse_layouts: bool,
    pub top_level_is_layout: bool,
    pub include_deleted: bool,
    pub inherited_state: Option<TraversalState>,
    pub skip_invisible_notes: bool,
}

impl Default for ConcatOpts {
    fn default() -> Self {
        Self {
            recurse_layouts: false,
            top_level_is_layout: true,
            include_deleted: false,
            inherited_state: None,
            skip_invisible_notes: false,
        }
    }
}

pub fn concatenate_text_nodes(
    doc: &Document,
    parent: NodeId,
    opts: &ConcatOpts,
) -> (Vec<TextSegment>, String) {
    let mut segments = Vec::new();
    let inherited = opts
        .inherited_state
        .clone()
        .unwrap_or_else(create_traversal_state);
    walk(
        doc,
        parent,
        opts.top_level_is_layout,
        &inherited,
        opts,
        &mut segments,
    );
    let full_text = segments.iter().map(|s| s.text.as_str()).collect::<String>();
    (segments, full_text)
}

fn walk(
    doc: &Document,
    list_owner: NodeId,
    collect_text: bool,
    inherited_state: &TraversalState,
    opts: &ConcatOpts,
    segments: &mut Vec<TextSegment>,
) {
    let ids = doc.node(list_owner).children.clone();
    let mut state = enter_traversal_state(inherited_state);
    for (i, &id) in ids.iter().enumerate() {
        match &doc.node(id).kind {
            NodeKind::Property { key, value } => {
                advance_traversal_state(&mut state, key, value.as_deref());
            }
            NodeKind::Text { text } => {
                if collect_text
                    && (opts.include_deleted || traversal_region(&state) != TextRegion::Deleted)
                {
                    let segment_state = clone_traversal_state(&state);
                    segments.push(TextSegment {
                        child_index: i,
                        text: text.clone(),
                        owner: if opts.recurse_layouts {
                            Some(list_owner)
                        } else {
                            None
                        },
                        state: segment_state,
                    });
                }
            }
            NodeKind::Block { tag, args, .. } if opts.recurse_layouts => {
                if tag == "layout" {
                    walk(doc, id, true, &state, opts, segments);
                } else if tag == "inset" {
                    if opts.skip_invisible_notes && is_invisible_inset(tag, args.as_deref()) {
                        continue;
                    }
                    walk(doc, id, false, &state, opts, segments);
                }
            }
            _ => {}
        }
    }
}

pub fn map_pos_to_segment(segments: &[TextSegment], pos: usize) -> (usize, usize) {
    let mut remaining = pos;
    for (i, seg) in segments.iter().enumerate() {
        if remaining < seg.text.len() {
            return (i, remaining);
        }
        remaining -= seg.text.len();
    }
    let last = segments
        .last()
        .expect("invariant: map_pos_to_segment on non-empty segments");
    (segments.len() - 1, last.text.len())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::parse;

    #[test]
    fn change_depths_zero_opposite() {
        assert_eq!(advance_change_depths("change_deleted", 0, 3), (1, 0));
        assert_eq!(advance_change_depths("change_inserted", 2, 0), (0, 1));
        assert_eq!(advance_change_depths("change_unchanged", 0, 1), (0, 0));
        assert_eq!(advance_change_depths("change_unchanged", 1, 0), (0, 0));
    }

    #[test]
    fn private_notes_only() {
        assert_eq!(
            invisible_inset_type("inset", Some("Note Note")),
            Some("Note")
        );
        assert_eq!(
            invisible_inset_type("inset", Some("Note Comment")),
            Some("Comment")
        );
        assert!(!is_invisible_inset("inset", Some("Note Greyedout")));
        assert!(!is_invisible_inset("layout", Some("Note Note")));
    }

    #[test]
    fn concatenate_skips_deleted_by_default() {
        let src = "\\begin_layout Standard\nkeep\n\\change_deleted 1 1\ngone\n\\change_unchanged\nmore\n\\end_layout";
        let doc = parse(src, true).unwrap();
        let layout = doc.node(doc.root()).children[0];
        let (_segs, full) = concatenate_text_nodes(&doc, layout, &ConcatOpts::default());
        assert_eq!(full, "keepmore");
        let (_segs, all) = concatenate_text_nodes(
            &doc,
            layout,
            &ConcatOpts {
                include_deleted: true,
                ..ConcatOpts::default()
            },
        );
        assert_eq!(all, "keepgonemore");
    }
}
